sw/banshee/src/tran.rs

```diff
@@ -1206,20 +1206,22 @@ impl<'a> InstructionTranslator<'a> {
 
     unsafe fn emit_imm12_rs1(&self, data: riscv::FormatImm12Rs1) -> Result<()> {
         let imm = data.imm();
-        trace!("{} x{}, x{}", data.op, data.rs1, imm);
+        trace!("{} x{}, x{:x}", data.op, data.rs1, imm);
 
         // Compute the address.
         let rs1 = self.read_reg(data.rs1);
-        let imm = LLVMConstInt(LLVMInt32Type(), (imm as i64) as u64, 0);
 
         // Perform the operation.
         match data.op {
             riscv::OpcodeImm12Rs1::Scfgwi => {
                 // ssr write immediate holds address offset in imm12, content in rs1
+                let dm = (imm as u64) & 0x1f;
+                let reg_word = ((imm as u64) >> 5) & 0x7f;
+                let addr_off = LLVMConstInt(LLVMInt32Type(), (dm << 8) | (reg_word << 3), 0);
                 let addr = LLVMBuildAdd(
                     self.builder,
                     LLVMConstInt(LLVMInt32Type(), SSR_BASE, 0),
-                    imm,
+                    addr_off,
                     NONAME,
                 );
                 self.write_mem(addr, rs1, 2);
```

Comment:
i am new to rust. If i add the line `_ => bail!("Unsupported opcode {}", data.op),` it  throws a warning about `unreachable pattern`. I guess in the current implementation, there is no other function matching `imm12_rd`, therefore when `emit_imm12_rd` is called, the opcode is **always** `Scfgri`.

Reply:
I am unsure whether `rustc` tracks matching across nested `match` constructs, but it may be. Since we are unsure in this case, it is best to add a comment to document this.

Reply:
we could also just leave it there and accept a compiler warning :) is probably safer, in case there is an instruction with this format in a  future version that is not yet parsed in `tran.rs`. What do you think @paulsc96 @SamuelRiedel ?

Reply:
Ah ok, that makes sense. I am against tolerating compiler warnings because that quickly leads to more and more warnings building up and ultimately, important warnings going unnoticed.

I propose add the `_ => bail ...` and disable the warning locally with `#![allow(unreachable_patterns)]`. If you write it as the first statement in the `match` clause it should disable the warning only locally. Coupled with a comment, this should prevent any future errors.

Reply:
done in 9020444. this can't be done unfortunately on the match statements that assign a value, `attributes on expressions are experimental`[issue 1570](https://github.com/rust-lang/rust/issues/15701)

Reply:
Nice :tada:. And very interesting to know that we can't do this on the `let match` statements. Thanks for looking into this.